File: y23_go/d08.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {
    // PREP
    // ----
    input := func() []string {
        input, err := os.ReadFile("d08.in");
        if err != nil {
            panic(err)
        }
        return strings.Split(string(input), "\n")
    }();

    // split the input between the L/R instructions and the network
    instructions, mappings := input[0], input[2:len(input)-1]

    // create the network mappings
    M := make(map[string][]string)
    for _, mapping := range mappings {
        node, node_mapping := func() (string, []string) {
            mapping := strings.Split(strings.ReplaceAll(mapping, " ", ""), "=")

            node := mapping[0]
            node_mapping := strings.Split(mapping[1][1:len(mapping[1])-1], ",")

            return node, node_mapping
        }();
        M[node] = node_mapping
    }

    {
        // PART ONE
        // --------
        // follow the path using the given instructions until it hits ZZZ
        start_node := "AAA"
        idx := 0
        step := 0
        for start_node != "ZZZ" {
            nodes := M[start_node]
            direction := instructions[idx]

            switch direction {
                case 'L': start_node = nodes[0]
                case 'R': start_node = nodes[1]
            }

            idx = (idx + 1) % len(instructions)
            step += 1
        }
        fmt.Printf("Part 1: %d\n", step)
    }


    {
        // PART TWO
        // --------
        type node_step struct {loc string; step int}

        // get starting nodes
        pairs := []node_step{}
        for node := range M {
            if node[2] == 'A' { pairs = append(pairs, node_step{loc: node, step: 0}) }
        }

        // check if all of the locations are at a 'Z' node
        check := func(pairs []node_step) bool {
            for _, pair := range pairs {
                if pair.loc[2] != 'Z' { return false }
            }
            return true
        }

        // for each location, follow the path using the given instructions until it is at a 'Z' node
        i := 0
        for !check(pairs) {
            for idx := range pairs {
                pair := &pairs[idx]
                if pair.loc[2] == 'Z' { continue }

                direction := instructions[i]
                switch direction {
                    case 'L': pair.loc = M[pair.loc][0]
                    case 'R': pair.loc = M[pair.loc][1]
                }
                pair.step += 1
            }
            i = (i + 1) % len(instructions)
        }

        // basic gcd/lcm implementation
        gcd := func(a, b int) int {
            for a != b {
                if a > b { a = a - b } else { b = b - a }
            }
            return a
        }

        lcd := func(a, b int) int {
            return a * b / gcd(a, b)
        }

        // lcm of all the total steps of each location is the answer
        res := pairs[0].step
        for _, pair := range pairs {
            res = lcd(res, pair.step)
        }
        fmt.Printf("Part 2: %d\n", res)
    }
}

File: y23_py/d08.py
from math import lcm

def partA(input: str):
    instructions, paths = input.split('\n\n')
    pathslist = paths.splitlines()
    
    paths = {}
    for path in pathslist:
        loc, options = path.split(' = ')
        options = options.replace('(', '').replace(')', '').split(', ')
        paths[loc] = options

    steps = 0
    current_location = 'AAA'
    i = 0
    while True:
        direction = instructions[i%len(instructions)]
        if current_location == 'ZZZ':
            return steps
        if direction == 'L':
            current_location = paths[current_location][0]
            steps += 1
            i += 1
            continue
        if direction == 'R':
            current_location = paths[current_location][1]
            steps += 1
            i += 1
            continue

def partB(input: str):
    instructions, paths = input.split('\n\n')
    pathslist = paths.splitlines()
    
    paths = {}
    for path in pathslist:
        loc, options = path.split(' = ')
        options = options.replace('(', '').replace(')', '').split(', ')
        paths[loc] = options

    locations = {}
    for path in paths:
        if path[2] == 'A':
            locations[path] = (path, 0)

    i = 0
    while True:
        direction = instructions[i%len(instructions)]
        for loc in locations:
            curr_loc, steps = locations[loc]
            if curr_loc[2] == 'Z':
                continue

            if direction == 'L':
                locations[loc] = (paths[curr_loc][0], steps+1)

            if direction == 'R':
                locations[loc] = (paths[curr_loc][1], steps+1)
        i += 1

        s = set([x[0][2] for x in locations.values()])
        if len(s) == 1 and 'Z' in s:
            return lcm(*[x[1] for x in locations.values()])

    
            

if __name__ == '__main__':
    import sys

    infile = sys.argv[1] if len(sys.argv) > 1 else 'd08.in'
    inp = open(infile).read().strip()
    print(f'A: {partA(inp[::])}')
    print(f'B: {partB(inp[::])}')
